integrations/gitlab2/gitlabcore/client.py

```diff
@@ -0,0 +1,253 @@
+from typing import Any, Optional, MutableMapping, Dict, AsyncGenerator, List
+import time
+import asyncio
+import httpx
+
+from loguru import logger
+
+from port_ocean.utils import http_async_client
+from port_ocean.utils.cache import cache_iterator_result
+
+RETRYABLE_ERROR_CODES = [500, 503]
+
+def is_retryable_status_code(status_code: Optional[int]) -> bool:
+    """
+    Checks if the status code is retryable.
+    Handles rate limiting (HTTP 429) and transient errors (e.g., HTTP 500, 503).
+    """
+    if status_code == 429:
+        return True
+
+    if status_code in RETRYABLE_ERROR_CODES:
+        return True
+
+    return False
+
+
+def calculate_wait_time(retries: int, headers: Optional[MutableMapping[str, str]] = None ) -> float:
+    """
+    Calculates the wait time before retrying a request.
+    For rate limits (HTTP 429), it respects the `Retry-After` or `RateLimit-Reset` headers.
+    Otherwise, it uses exponential backoff.
+    """
+    wait_time = 2**retries * 0.1 # Exponential backoff base time
+
+    # https://docs.gitlab.com/ee/administration/settings/user_and_ip_rate_limits.html#response-headers
+    if headers:
+        if "Retry-After" in headers:
+            wait_time = int(headers["Retry-After"])
+        elif "RateLimit-Reset" in headers:
+            wait_time = int(headers["RateLimit-Reset"]) - time.time()
+
+    return max(wait_time, 0.1) # Ensure the wait time is at least 0.1 seconds
+
+async def handle_retry_logic(status_code: Optional[int], headers: Optional[MutableMapping[str, str]], retries: int = 0, max_retries: int = 5) -> bool:
+    """
+    Handles the retry logic based on the status code and retry count.
+    - If the status code is retryable (rate limit or transient error), it calculates the wait time.
+    - Waits for the calculated time before allowing the retry.
+    - Stops retrying if the maximm retry count is exceeded.
+    """
+    if not is_retryable_status_code(status_code) or retries >= max_retries :
+        return False
+
+    wait_time = calculate_wait_time(retries, headers)
+    retries += 1
+
+    logger.info(f"Retrying after {wait_time} seconds (attempt {retries})...")
+    await asyncio.sleep(wait_time)
+
+    return True
+
+
+# PaginatedData Class
+class PaginatedData:
```

Comment:
You should be able to handle pagination using a method in the client class. While this is a clever idea, it is good to keep things as simple as possible